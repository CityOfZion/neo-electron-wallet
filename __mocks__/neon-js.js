const { Fixed8 } = require('neon-js').u

const neonjs = jest.genMockFromModule('neon-js')

const promiseMockGen = (result, error = false) => {
  return jest.fn(() => {
    return new Promise((resolve, reject) => {
      if (error) reject(error)
      resolve(result)
    })
  })
}

const privateKey = 'L4AJ14CNaBWPemRJKC34wyZwbmxg33GETs4Y1F8uK7rRmZ2UHrJn'
const address = 'AM22coFfbe9N6omgL9ucFBLkeaMNg9TEyL'
const encryptedKey = '6PYUGtvXiT5TBetgWf77QyAFidQj61V8FJeFBFtYttmsSxcbmP4vCFRCWu'
const scriptHash = '4bcdc110b6514312ead9420467475232d4f08539'

neonjs.api = {
  neonDB: {
    getMaxClaimAmount: promiseMockGen(new Fixed8('1.59140785')),
    getClaims: promiseMockGen({
      claims: [
        { claim: new Fixed8('1.28045113') },
        { claim: new Fixed8('0.31095672') }
      ]
    }),
    doClaimAllGas: promiseMockGen({ result: true }),
    doSendAsset: promiseMockGen({ result: true }),
    getWalletDBHeight: promiseMockGen(586435),
    getAPIEndpoint: jest.fn(() => 'http://testnet-api.wallet.cityofzion.io'),
    getRPCEndpoint: promiseMockGen(''),
    doMintTokens: promiseMockGen({ result: true }),
    getTransactionHistory: promiseMockGen([]),
    getBalance: promiseMockGen({
      assets: {
        NEO: { balance: new Fixed8(1) },
        GAS: { balance: new Fixed8(1) }
      }
    })
  },
  nep5: {
    getTokenInfo: promiseMockGen({ result: true }),
    getTokenBalance: promiseMockGen(100)
  }
}

neonjs.create = {
  account: {
    address
    // privateKey
    // wif
  }
}
neonjs.tx = {
  serializeTransaction: jest.fn()
}
// TODO - look into why I chose to use encrypt vs encryptWif from new API
neonjs.wallet = {
  getPublicKeyEncoded: jest.fn(),
  decrypt: jest.fn(() => privateKey),
  encrypt: jest.fn(() => encryptedKey),
  generatePrivateKey: jest.fn(() => privateKey),
  Account: jest.fn(() => { return { address } }),
  getVerificationScriptFromPublicKey: jest.fn(() => scriptHash),
  isAddress: jest.fn(() => true),
  isNEP2: jest.fn(() => true)
}

module.exports = neonjs
