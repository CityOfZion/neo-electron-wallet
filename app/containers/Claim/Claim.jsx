// @flow
import React, { Component } from 'react'

import Tooltip from '../../components/Tooltip'
import { formatGAS } from '../../core/formatters'

type Props = {
  doClaimNotify: Function,
  setClaimRequest: Function,
  doGasClaim: Function,
  claimRequest: boolean,
  disableClaimButton: boolean,
  claimWasUpdated: boolean,
  claimAmount: number,
}

export default class Claim extends Component<Props> {
  componentDidUpdate () {
    const { claimRequest, claimWasUpdated, doClaimNotify, setClaimRequest } = this.props
    if (claimRequest && claimWasUpdated) {
      setClaimRequest(false)
      doClaimNotify()
    }
  }

  render () {
    const { claimAmount, disableClaimButton, doGasClaim } = this.props
    const shouldDisableButton = disableClaimButton || claimAmount === 0
    const formattedAmount = formatGAS(claimAmount)
    return (
      <div id='claim'>
        <Tooltip
          title='You can claim GAS once every 5 minutes'
          disabled={!disableClaimButton}
        >
          <button
            disabled={shouldDisableButton}
            onClick={() => doGasClaim()}
            className={shouldDisableButton ? 'disabled' : ''}>Claim {formattedAmount} GAS</button>
        </Tooltip>
      </div>
    )
  }
}
