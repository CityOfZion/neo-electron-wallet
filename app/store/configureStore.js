import { createStore, applyMiddleware } from 'redux';
import { createLogger } from 'redux-logger';
import thunk from 'redux-thunk';
import rootReducer from '../reducers/rootReducer';
// export type GlobalState = {}
export var INITIAL_STATE = {};
var loggerMiddleware = createLogger();
function configureStore(initialState) {
    if (initialState === void 0) { initialState = INITIAL_STATE; }
    return createStore(rootReducer, initialState, process.env.NODE_ENV === 'production'
        ? applyMiddleware(thunk)
        : applyMiddleware(thunk, loggerMiddleware));
}
export var store = configureStore();
