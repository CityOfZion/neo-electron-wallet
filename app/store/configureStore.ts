import { createStore, applyMiddleware } from 'redux'
import { createLogger } from 'redux-logger'
import thunk from 'redux-thunk'
import rootReducer from '../reducers/rootReducer'

// export type GlobalState = {}
export const INITIAL_STATE = {}

const loggerMiddleware = createLogger()

function configureStore(initialState = INITIAL_STATE) {
  return createStore(
    rootReducer,
    initialState,
    process.env.NODE_ENV === 'production'
      ? applyMiddleware(thunk)
      : applyMiddleware(thunk, loggerMiddleware),
  )
}
export const store = configureStore()
