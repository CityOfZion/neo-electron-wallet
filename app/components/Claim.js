import React, { Component } from 'react';
import { connect } from 'react-redux';
import { doClaimAllGas, doSendAsset } from 'neon-js';
import { sendEvent, clearTransactionEvent, setClaimRequest, disableClaim } from '../actions/index.js';
import ReactTooltip from 'react-tooltip'

// To initiate claim, first send all Neo to own address, the set claimRequest state
// When new claims are available, this will trigger the claim
const doGasClaim = (dispatch, net, wif, selfAddress, ans) => {
  dispatch(sendEvent(true, "Sending Neo to Yourself..."));
  doSendAsset(net, selfAddress, wif, "Neo", ans).then((response) => {
    if (response.result === undefined){
      dispatch(sendEvent(false, "Transaction failed!"));
    } else {
      dispatch(sendEvent(true, "Waiting for transaction to clear..."));
      dispatch(setClaimRequest(true));
      dispatch(disableClaim(true));
    }
  });
}

class Claim extends Component {

  componentDidUpdate = () => {
    // if we requested a claim and new claims are available, do claim
    if (this.props.claimRequest === true && this.props.claimWasUpdated == true){
      this.props.dispatch(setClaimRequest(false));
      doClaimAllGas(this.props.net, this.props.wif).then((response) => {
        if (response.result === true){
          this.props.dispatch(sendEvent(true, "Claim was successful! Your balance will update once the blockchain has processed it."));
          setTimeout(() => this.props.dispatch(disableClaim(false)), 300000);
        } else {
          this.props.dispatch(sendEvent(false, "Claim failed"))
        }
        setTimeout(() => this.props.dispatch(clearTransactionEvent()), 5000);
      });
    }
  }

  render = () => {
    let renderButton;
    const doClaim = () => doGasClaim(this.props.dispatch, this.props.net, this.props.wif, this.props.address, this.props.neo);
    if (this.props.disableClaimButton === false){
      renderButton = <button onClick={doClaim}>Claim {this.props.claimAmount} GAS</button>;
    } else {
      renderButton = (<div>
          <button data-tip data-for="claimTip" className="disabled">Claim {this.props.claimAmount} GAS</button>
          <ReactTooltip class="solidTip" id="claimTip" place="bottom" type="dark" effect="solid">
            <span>You can claim Gas once every 5 minutes</span>
          </ReactTooltip>
        </div>);
    }
    return <div id="claim">{renderButton}</div>;
  }

}

const mapStateToProps = (state) => ({
  claimAmount: state.claimState.claimAmount,
  claimRequest: state.claimState.claimRequest,
  claimWasUpdated: state.claimState.claimWasUpdated,
  disableClaimButton: state.claimState.disableClaimButton,
  wif: state.account.wif,
  address: state.account.address,
  net: state.metadata.network,
  neo: state.wallet.Neo
});

Claim = connect(mapStateToProps)(Claim);

export default Claim;
