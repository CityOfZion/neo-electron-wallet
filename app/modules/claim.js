// @flow
import { api, rpc } from 'neon-js'

import {
  showErrorNotification,
  showSuccessNotification,
  showInfoNotification
} from './notifications'
import {
  getNetwork,
  getWIF,
  getAddress,
  getSigningFunction,
  getPublicKey,
  getIsHardwareLogin,
  getNEO
} from '../core/deprecated'
import { ASSETS } from '../core/constants'
import { FIVE_MINUTES_MS } from '../core/time'

import { log } from '../util/Logs'

// Constants
export const SET_CLAIM_REQUEST = 'SET_CLAIM_REQUEST'
export const DISABLE_CLAIM = 'DISABLE_CLAIM'
export const SET_FINALIZE_CLAIM = 'SET_FINALIZE_CLAIM'
export const SET_TRANSACTION_ID = 'SET_TRANSACTION_ID'

// Actions
export function setClaimRequest (claimRequest: boolean) {
  return {
    type: SET_CLAIM_REQUEST,
    payload: { claimRequest }
  }
}
export function setTransactionId (transactionId: string) {
  return {
    type: SET_TRANSACTION_ID,
    payload: { transactionId }
  }
}

export function setFinalizeClaim (finalizeClaim: boolean) {
  return {
    type: SET_FINALIZE_CLAIM,
    payload: { finalizeClaim }
  }
}

export function disableClaim (disableClaimButton: boolean) {
  return {
    type: DISABLE_CLAIM,
    payload: { disableClaimButton }
  }
}

export const doClaimNotify = () => async (
  dispatch: DispatchType,
  getState: GetStateType
) => {
  const state = getState()
  const address = getAddress(state)
  const net = getNetwork(state)
  const signingFunction = getSigningFunction(state)
  const publicKey = getPublicKey(state)
  const privateKey = getWIF(state)
  const isHardwareClaim = getIsHardwareLogin(state)

  log(net, 'CLAIM', address, { info: 'claim all GAS' })

  if (isHardwareClaim) {
    dispatch(showInfoNotification({
      message: 'Sign transaction 2 of 2 to claim GAS on your hardware device (claiming GAS)',
      autoDismiss: 0
    }))
  }

  try {
    const { response } = await api.claimGas({ net, address, publicKey, privateKey, signingFunction })

    if (!response.result) {
      throw new Error('Claim failed')
    }

    dispatch(showSuccessNotification({
      message: 'Claim was successful! Your balance will update once the blockchain has processed it.'
    }))
    setTimeout(() => dispatch(disableClaim(false)), FIVE_MINUTES_MS)
  } catch (err) {
    return dispatch(showErrorNotification({ message: 'Claim failed' }))
  }
}

// To initiate claim, first send all NEO to own address, the set claimRequest state
// When new claims are available, this will trigger the claim
export const doGasClaim = () => async (
  dispatch: DispatchType,
  getState: GetStateType
) => {
  const state = getState()
  const address = getAddress(state)
  const net = getNetwork(state)
  const NEO = getNEO(state)
  const publicKey = getPublicKey(state)
  const privateKey = getWIF(state)
  const signingFunction = getSigningFunction(state)
  const isHardwareClaim = getIsHardwareLogin(state)

  // if no NEO in account, no need to send to self first
  if (NEO === '0') {
    return dispatch(doClaimNotify())
  } else {
    dispatch(showInfoNotification({
      message: 'Sending NEO to Yourself...',
      autoDismiss: 0
    }))
    log(net, 'SEND', address, { to: address, amount: NEO, asset: ASSETS.NEO })

    if (isHardwareClaim) {
      dispatch(showInfoNotification({
        message: 'Sign transaction 1 of 2 to claim GAS on your hardware device (sending NEO to yourself)',
        autoDismiss: 0
      }))
    }

    try {
      const { response } = await api.sendAsset({
        net,
        address,
        publicKey,
        privateKey,
        signingFunction,
        intents: api.makeIntent({ [ASSETS.NEO]: NEO }, address)
      })

      if (!response.result || !response.txid) {
        throw new Error('Transaction failed!')
      }

      dispatch(showInfoNotification({
        message: 'Waiting for transaction to clear...',
        autoDismiss: 0
      }))
      dispatch(setTransactionId(response.txid))
      dispatch(setClaimRequest(true))
      return dispatch(disableClaim(true))
    } catch (err) {
      return dispatch(showErrorNotification({ message: 'Transaction failed!' }))
    }
  }
}

export const checkClaimStatus = () => async (
  dispatch: DispatchType,
  getState: GetStateType
) => {
  const state = getState()
  const net = getNetwork(state)
  const txid = getTransactionId(state)
  const endpoint = await api.loadBalance(api.getRPCEndpointFrom, { net })
  const response = await rpc.Query.getRawTransaction(txid).execute(endpoint)
  if (response.result && response.result.blockhash && response.result.confirmations) {
    dispatch(setTransactionId(''))
    dispatch(setFinalizeClaim(true))
  }
}

// State Getters
export const getFinalizeClaim = (state: Object) => state.claim.finalizeClaim
export const getClaimRequest = (state: Object) => state.claim.claimRequest
export const getDisableClaimButton = (state: Object) => state.claim.disableClaimButton
export const getTransactionId = (state: Object) => state.claim.transactionId

const initialState = {
  claimRequest: false,
  disableClaimButton: false,
  transactionId: '',
  finalizeClaim: false
}

export default (state: Object = initialState, action: ReduxAction) => {
  switch (action.type) {
    case SET_TRANSACTION_ID:
      const { transactionId } = action.payload
      return {
        ...state,
        transactionId
      }
    case SET_FINALIZE_CLAIM:
      const { finalizeClaim } = action.payload
      return {
        ...state,
        finalizeClaim
      }
    case SET_CLAIM_REQUEST:
      const { claimRequest } = action.payload
      return {
        ...state,
        claimRequest
      }
    case DISABLE_CLAIM:
      const { disableClaimButton } = action.payload
      return {
        ...state,
        disableClaimButton
      }
    default:
      return state
  }
}
