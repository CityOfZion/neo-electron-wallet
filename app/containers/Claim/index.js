// @flow
import { connect } from 'react-redux'
import { bindActionCreators } from 'redux'
import { compose } from 'recompose'

import Claim from './Claim'
import claimsActions from '../../actions/claimsActions'
import withData from '../../hocs/api/withData'
import {
  setClaimRequest,
  doGasClaim,
  doClaimNotify,
  getClaimRequest,
  getDisableClaimButton,
  getFinalizeClaim,
  checkClaimStatus
} from '../../modules/claim'

const mapStateToProps = (state: Object) => ({
  finalizeClaim: getFinalizeClaim(state),
  claimRequest: getClaimRequest(state),
  disableClaimButton: getDisableClaimButton(state)
})

const actionCreators = {
  setClaimRequest,
  doGasClaim,
  doClaimNotify,
  checkClaimStatus
}

const mapDispatchToProps = dispatch => bindActionCreators(actionCreators, dispatch)

const mapClaimsDataToProps = (claims) => ({
  claimAmount: claims.total
})

export default compose(
  connect(mapStateToProps, mapDispatchToProps),
  withData(claimsActions, mapClaimsDataToProps)
)(Claim)
