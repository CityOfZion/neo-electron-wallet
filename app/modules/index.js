// @flow
import { combineReducers } from 'redux'
import api from './api'
import generateWallet from './generateWallet'
import transactions from './transactions'
import wallet from './wallet'
import claim from './claim'
import notifications from './notifications'
import modal from './modal'
import addressBook from './addressBook'

export default combineReducers({
  api,
  generateWallet,
  wallet,
  transactions,
  claim,
  notifications,
  modal,
  addressBook
})
