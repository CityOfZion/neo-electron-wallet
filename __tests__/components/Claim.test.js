import React from 'react'
import { cloneDeep } from 'lodash'
import { Provider } from 'react-redux'
import configureStore from 'redux-mock-store'
import thunk from 'redux-thunk'
import { shallow, mount } from 'enzyme'
import Claim from '../../app/containers/Claim'
import * as neonjs from 'neon-js'
import { setClaimRequest, disableClaim } from '../../app/modules/claim'

const initialState = {
  claim: {
    claimAmount: 10,
    claimRequest: false,
    claimWasUpdated: false,
    disableClaimButton: false
  },
  account: {
    wif: 'wif',
    address: 'address'
  },
  metadata: {
    network: 'network'
  },
  wallet: {
    Neo: 1
  }
}

const setup = (state = initialState, shallowRender = true) => {
  const store = configureStore([thunk])(state)

  let wrapper
  if (shallowRender) {
    wrapper = shallow(<Claim store={store} />)
  } else {
    wrapper = mount(
      <Provider store={store}>
        <Claim />
      </Provider>
    )
  }

  return {
    store,
    wrapper
  }
}

describe('Claim', () => {
  test('should render without crashing', () => {
    const { wrapper } = setup()
    expect(wrapper).toMatchSnapshot()
  })

  test('should render claim gas button when claim button is not disabled', () => {
    const { wrapper } = setup()
    expect(wrapper.dive()).toMatchSnapshot()
  })

  test('should not render claim gas button when claim button is disabled', () => {
    const newState = cloneDeep(initialState)
    newState.claim.disableClaimButton = true
    const { wrapper } = setup(newState)
    expect(wrapper.dive()).toMatchSnapshot()
  })

  describe('when do gas claim button is clicked', () => {
    test('should dispatch transaction failure event', async (done) => {
      const { wrapper, store } = setup()
      neonjs.doSendAsset = jest.fn(() => {
        return new Promise((resolve, reject) => {
          resolve({ result: undefined })
        })
      })
      wrapper.dive().find('#claim button').simulate('click')

      await Promise.resolve().then().then().then()
      const actions = store.getActions()
      expect(actions.length).toEqual(2)
      expect(actions[0]).toEqual({
        payload: {
          message: 'Sending Neo to Yourself...',
          type: 'INFO'
        },
        type: 'SHOW_NOTIFICATION'
      })
      expect(actions[1]).toEqual({
        payload: {
          message: 'Transaction failed!',
          type: 'ERROR'
        },
        type: 'SHOW_NOTIFICATION'
      })
      done()
    })

    test('should dispatch transaction waiting, set claim request and disable claim event', async (done) => {
      const { wrapper, store } = setup()
      neonjs.doSendAsset = jest.fn(() => {
        return new Promise((resolve, reject) => {
          resolve({ result: true })
        })
      })

      wrapper.dive().find('#claim button').simulate('click')

      await Promise.resolve().then().then().then()
      const actions = store.getActions()
      expect(actions.length).toEqual(4)
      expect(actions[0]).toEqual({
        payload: {
          message: 'Sending Neo to Yourself...',
          type: 'INFO'
        },
        type: 'SHOW_NOTIFICATION'
      })
      expect(actions[1]).toEqual({
        payload: {
          message: 'Waiting for transaction to clear...',
          type: 'INFO'
        },
        type: 'SHOW_NOTIFICATION'
      })
      expect(actions[2]).toEqual(setClaimRequest(true))
      expect(actions[3]).toEqual(disableClaim(true))
      done()
    })
  })
})
