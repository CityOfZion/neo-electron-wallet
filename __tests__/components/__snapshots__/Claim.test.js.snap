// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Claim should not render claim GAS button when claim button is disabled 1`] = `
<div
  id="claim"
>
  <Tooltip
    disabled={false}
    title="You can claim GAS once every 5 minutes"
  >
    <button
      className="disabled"
      disabled={true}
      onClick={[Function]}
    >
      Claim 
      10.00000000
       GAS
    </button>
  </Tooltip>
</div>
`;

exports[`Claim should render claim GAS button when claim button is not disabled 1`] = `
<div
  id="claim"
>
  <Tooltip
    disabled={true}
    title="You can claim GAS once every 5 minutes"
  >
    <button
      className=""
      disabled={false}
      onClick={[Function]}
    >
      Claim 
      10.00000000
       GAS
    </button>
  </Tooltip>
</div>
`;

exports[`Claim should render without crashing 1`] = `
<Claim
  claimAmount={10}
  claimRequest={false}
  claimWasUpdated={false}
  disableClaimButton={false}
  doClaimNotify={[Function]}
  doGasClaim={[Function]}
  setClaimRequest={[Function]}
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
  storeSubscription={
    Subscription {
      "listeners": Object {
        "clear": [Function],
        "get": [Function],
        "notify": [Function],
        "subscribe": [Function],
      },
      "onStateChange": [Function],
      "parentSub": undefined,
      "store": Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      },
      "unsubscribe": [Function],
    }
  }
/>
`;
